File: k4sdl/sprite/SpriteLoader.cpp
#include "SpriteLoader.hpp"

namespace k4sdl {
    void SpriteLoaderOS::load(std::string path, Sprite &sprite,
                              bool spriteSheet, SDL_Color* colorKey) {
        std::string realPath = basePath + path;
        std::vector<Frame> frameInfo;
        std::vector<Tag> tagInfo;
        SDL_Texture* texture;

        if (spriteSheet) {
            std::string dataFile = realPath + ".dat";
            std::ifstream f(dataFile.c_str(), std::ios::binary);
            if (!f.good()) {
                std::cout << "Error opening spritesheet data. Path: " << dataFile << std::endl;
                return;
            }

            char header[4];
            f.read(header, 4);

            if (memcmp(header, "SPR\x20", 4) != 0) {
                f.close();
                std::cout << "Error on spritesheet data header. Path: " << dataFile << std::endl;
                return;
            }

            uint8_t frameCount, tagCount;
            f.read((char*)&frameCount, 1);
            f.read((char*)&tagCount, 1);

            frameInfo.resize(frameCount);
            for (int i = 0; i < frameCount; i++) {
                f.read((char*)&frameInfo[i].x, 2);
                f.read((char*)&frameInfo[i].y, 2);
                f.read((char*)&frameInfo[i].w, 2);
                f.read((char*)&frameInfo[i].h, 2);
            }

            tagInfo.resize(tagCount);
            for (int i = 0; i < tagCount; i++) {
                f.read((char*)&tagInfo[i].childXOff, 2);
                f.read((char*)&tagInfo[i].childYOff, 2);
                f.read((char*)&tagInfo[i].childTagNum, 2);
                std::getline(f, tagInfo[i].name, '\0');

                uint8_t tagFrameCount;
                f.read((char*)&tagFrameCount, 1);

                tagInfo[i].frames.resize(tagFrameCount);
                for (int j = 0; j < tagFrameCount; j++) {
                    f.read((char*)&tagInfo[i].frames[j], 1);
                }

                tagInfo[i].frameDurations.resize(tagFrameCount);
                for (int j = 0; j < tagFrameCount; j++) {
                    f.read((char*)&tagInfo[i].frameDurations[j], 4);
                }
            }

            f.close();
        }

        SDL_Surface* loadedSurface = IMG_Load(realPath.c_str());
        if (loadedSurface == NULL) {
            std::cout << "Error loading surface. Path: " << realPath << " Error: " << IMG_GetError() << std::endl;
            return;
        }

        if (colorKey != NULL)
            SDL_SetColorKey(loadedSurface, SDL_TRUE, SDL_MapRGB(loadedSurface->format, colorKey->r, colorKey->g, colorKey->b));
        
        texture = SDL_CreateTextureFromSurface(Screen::getRenderer(), loadedSurface);
        SDL_FreeSurface(loadedSurface);
        if (texture == NULL) {
            std::cout << "Error creating texture. Path: " << realPath << " Error: " << SDL_GetError() << std::endl;
            return;
        }

        sprite.loadSprite(texture, frameInfo, tagInfo);
    }
}

File: tools/contruct_spritesheet/pg_engine/sprite/SpriteLoader.py
import os
from .Sprite import Tag, Frame
import json


class SpriteLoader:
    def load(self, path: str):
        pass


class SpriteLoaderOS(SpriteLoader):
    def __init__(self, base_path_os=None):
        self.base_path = base_path_os

    def load(self, path: str):
        if self.base_path:
            path = os.path.join(self.base_path, path)

        if not os.path.isfile(path):
            return None, None
            
        frame_info = []
        tag_info = []
        with open(path, "r") as json_file:
            data = json.loads(json_file.read())

        for frame in data["frames"]:
            frame_data = frame["frame"]
            frame_info.append(Frame(frame_data["x"], frame_data["y"], frame_data["w"], frame_data["h"]))

        for tag in data["meta"]["frameTags"]:
            frames = []
            frame_durations = []
            if "frames" in tag:
                frames = tag["frames"]
            elif "from" in tag and "to" in tag:
                frames = list(range(tag["from"], tag["to"] + 1))
            if "durations" in tag:
                for duration in tag["durations"]:
                    frame_durations.append(duration / 1000.0)
            else:
                for frame in frames:
                    frame_durations.append(data["frames"][frame]["duration"] / 1000.0)
            tag_info.append(Tag(tag["name"], frames, frame_durations, 0, 0, 0))
        
        return frame_info, tag_info
